### PR TITLE
Adds support for nullable casts and missable/missingCall functions

### LINKED ISSUE #1398: Confirm plan typer behavior for equality between a type with a missable type
Currently the `PlanTyper` will determine that calling `=` with a `type` and `union(type, missing)` will be a dynamic function call of `EQ__TYPE_TYPE__BOOL` and `EQ__ANY_ANY__BOOL`.

This can be seen by the below case:
```
-- below creates a dynamic call
CAST(1 AS INT4) = CAST(`1e0` as INT4)
```
, which outputs a `dynamic` rex op in the output plan.
```
⚬ PartiQLPlan[tag=Plan-1b087f8c]
   └──Query[tag=Plan-c60300c1]
      └──Rex[type=union(bool, null), tag=Plan-f83abf85]
         └──Dynamic[tag=Plan-f1007872]
            ├──Rex[type=int4, tag=Plan-f80ac24c]
            │  └──Static[tag=Plan-1abf6ded]
            │     ├──Fn[signature=CAST_INT32__INT32__INT32, tag=Plan-70959201]
            │     └──Rex[type=int4, tag=Plan-938c10e8]
            │        └──Lit[value=Int32ValueImpl(value=1, annotations=[]), tag=Plan-50664d78]
            ├──Rex[type=union(int4, missing), tag=Plan-5cd108a4]
            │  └──Static[tag=Plan-b6e6c73f]
            │     ├──Fn[signature=CAST_INT32__FLOAT64__INT32, tag=Plan-b28ae936]
            │     └──Rex[type=float, tag=Plan-9a8f3175]
            │        └──Lit[value=Float64ValueImpl(value=1.0, annotations=[]), tag=Plan-f3b0eb1e]
            ├──Candidate[tag=Plan-cf153ca8]
            │  └──Fn[signature=EQ__INT32_INT32__BOOL, tag=Plan-d7b0c1e]
            └──Candidate[tag=Plan-e40c23c3]
               └──Fn[signature=EQ__ANY_ANY__BOOL, tag=Plan-1d699bf6]
```

## Additional Context
- Java version: 11
- PartiQL version: 0.14.5-SNAPSHOT

### PR DESCRIPTION
**Pull request by johnedquinn**
## Relevant Issues
- Closes #1398 

## Description
- Adds support for nullable casts and missable/missingCall functions. Since branch `v1` largely has these changes already, this is a port of its functionality.
- Adds a dedicated test file to test static vs dynamic invocation of functions. Included within this test file are tests pertaining to #1398 as well as other edge-cases.

## Other Information
- Updated Unreleased Section in CHANGELOG: **YES**
- Any backward-incompatible changes? **NO**
- Any new external dependencies? **NO**
- Do your changes comply with the [Contributing Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md)
  and [Code Style Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md)? **YES**

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt [hunk 1]
```diff
@@ -584,18 +584,21 @@ internal class PlanTyper(
 
             // Type the arguments
```
Comment (RCHowell) [inline partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt:585]: I'm curious if https://github.com/partiql/partiql-lang-kotlin/blob/v1/partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt#L1357-L1405 can be a replacement for this? This is a lot of logic that isn't in V1. Ideally we can bring the latest to 0.14.+

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/TypeLattice.kt [hunk 2]
```diff
@@ -176,7 +176,7 @@ internal class TypeLattice private constructor(
                 ANY to coercion()
             )
             graph[NULL] = relationships(
-                NULL to coercion()
+                *types.map { it to coercion() }.toTypedArray()
```
Comment (RCHowell) [inline partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/TypeLattice.kt:179]: nice

File: partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/operator/OpConcatTest.kt [hunk 3]
```diff
@@ -31,9 +32,20 @@ class OpConcatTest : PartiQLTyperTestBase() {
...
+                    return@forEach
+                }
+                if (args.all { it is NullType }) {
+                    val nullableHighestPrecedenceType = StaticType.unionOf(StaticType.STRING, StaticType.NULL)
+                    (this[TestResult.Success(nullableHighestPrecedenceType)] ?: setOf(args)).let {
+                        put(TestResult.Success(nullableHighestPrecedenceType), it + setOf(args))
+                    }
+                } else if (args.contains(StaticType.NULL)) {
+                    (this[TestResult.Success(StaticType.unionOf(args.toSet()))] ?: setOf(args)).let {
+                        put(TestResult.Success(StaticType.unionOf(args.toSet())), it + setOf(args))
```
Comment (RCHowell) [inline partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/operator/OpConcatTest.kt:48]: I had to update a lot of tests in V1. Could you please pull those to main with this latest update. My general ask here (sorry if it's a pain) is that V1 work has addressed some fn typing bugs and you've perhaps re-implemented the fix into 0.14. Hopefully we can port the specific parts of V1 to here.

https://github.com/partiql/partiql-lang-kotlin/commit/324926dac95a4946e5c81d59f4c77a5ed5abf482#diff-6aef277927cb088c80994b061cba107d0485ae8acfc82134225788841aa8a852

**Review by RCHowell — CHANGES_REQUESTED** [review]

File: partiql-types/src/main/kotlin/org/partiql/types/function/FunctionSignature.kt [hunk 4]
```diff
@@ -57,6 +60,8 @@ public sealed class FunctionSignature(
         isNullable: Boolean = true,
         @JvmField public val isDeterministic: Boolean = true,
         @JvmField public val isNullCall: Boolean = false,
+        @JvmField public val isMissable: Boolean = false,
```
[the remark this answers is not in the document]
Comment (alancai98) [reply partiql-types/src/main/kotlin/org/partiql/types/function/FunctionSignature.kt:63]: A little confused w/ the divergence in behavior. Seems like all the functions defined in the `PartiQLHeader` are all not "missible"?

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/PartiQLHeader.kt [hunk 5]
```diff
@@ -182,15 +184,24 @@ internal object PartiQLHeader : Header() {
         unary("neg", t, t)
     }
 
-    private fun eq(): List<FunctionSignature.Scalar> = types.all.map { t ->
+    private fun eq(): List<FunctionSignature.Scalar> = types.all.filterNot { it == ANY || it == MISSING }.map { t ->
```
Comment (alancai98) [inline partiql-planner/src/main/kotlin/org/partiql/planner/internal/PartiQLHeader.kt:187]: There's a divergence here between the modeling of functions between this PR and the `v1` branch. E.g.
- For `eq`, the `v1` branch includes all the types and they have the same `isNullable, `isMissingCall`, etc. configurations https://github.com/partiql/partiql-lang-kotlin/blob/2879f3a0a1d3197cc00fda6f1613490f88d19353/partiql-spi/src/test/kotlin/org/partiql/spi/connector/sql/SqlHeader.kt#L233-L241
- For the logical ops, the `v1` branch uses the defaults for `isMissable = true` and `isMissingCall = true`, whereas this PR defines `isMissable = false` and `isMissingCall = false`

I'm curious the reasoning for the modeling differences.

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt (hunk 1 above)
Comment (alancai98) [reply partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt:585]: +1 to Robert's question. Can the same logic from `v1` be ported to this PR/`main`? Have a couple concerns that
- the PR may have a different typing behavior from the `v1` branch
- the code changes may make it tricky to merge `main` into `v1` (or `v1` into `main`)

**Review by alancai98 — COMMENTED** [review]
Haven't looked at the tests in too much detail. Raising some initial concerns + questions in this review.

**Comment by johnedquinn** [comment]
Closing as an alternative solution is being actively pursued.
